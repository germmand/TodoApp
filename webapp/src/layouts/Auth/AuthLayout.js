// React
import React from 'react';
import PropTypes from 'prop-types';
import { Switch, Route, Redirect } from 'react-router-dom';

// Material-UI
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

// Redux
import { connect } from 'react-redux';

// Auth Views
import authRoutes from '../../routes/Auth';

const authViews = (
    <Switch>
        {authRoutes.map((prop) => {
          if (prop.redirect) {
            return <Redirect from={prop.path} to={prop.to} key={prop.key} />;
          }
          return <Route path={prop.path} component={prop.component} key={prop.key} />;
        })}
    </Switch>
);

const styles = {
  grow: {
    flexGrow: 1,
  },
  viewContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
    height: '90vh',
  },
};

class AuthLayout extends React.Component {
    onHandleLoginRoute = () => {
      const { history, location } = this.props;
      if (location.pathname !== '/auth/login') {
        history.push('/auth/login');
      }
    };

    onHandleRegisterRoute = () => {
      const { history, location } = this.props;
      if (location.pathname !== '/auth/signup') {
        history.push('/auth/signup');
      }
    };

    onDashboardRedirect = () => {
      const { history, isLoggedIn } = this.props;
      if (isLoggedIn) {
        history.replace('/home/Dashboard');
      }
    }

    componentDidMount = () => {
      // Redirects to dashboard in case
      // the user is already logged in
      // on opening/rendering.
      this.onDashboardRedirect();
    }

    componentDidUpdate = () => {
      // Redirects to dashboard in case
      // the user is already logged in
      // on action.
      this.onDashboardRedirect();
    }

    render() {
      const { classes } = this.props;

      return (
        <div>
            <AppBar position="static">
                <Toolbar>
                    <Typography variant="h6" color="inherit" className={classes.grow}>
                        TodoApp - Welcome!
                    </Typography>
                    <Button variant="outlined" color="inherit" onClick={this.onHandleLoginRoute}>
                        Login
                    </Button>
                    <Button color="inherit" onClick={this.onHandleRegisterRoute}>
                        Register
                    </Button>
                </Toolbar>
            </AppBar>
            <div className={classes.viewContainer}>
                {authViews}
            </div>
        </div>
      );
    }
}

AuthLayout.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  location: PropTypes.objectOf(PropTypes.string).isRequired,
  history: PropTypes.shape({
    action: PropTypes.string.isRequired,
    block: PropTypes.func.isRequired,
    createHref: PropTypes.func.isRequired,
    go: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
    goForward: PropTypes.func.isRequired,
    length: PropTypes.number.isRequired,
    listen: PropTypes.func.isRequired,
    location: PropTypes.objectOf(PropTypes.string).isRequired,
    push: PropTypes.func.isRequired,
    replace: PropTypes.func.isRequired,
  }).isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  isLoggedIn: state.authenticationReducer.isLoggedIn,
});

export default connect(
  mapStateToProps,
  null,
)(withStyles(styles)(AuthLayout));
