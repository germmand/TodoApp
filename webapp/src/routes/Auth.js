import AuthViews from '../views/Auth';

const authRoutes = [
  {
    path: '/auth/login',
    component: AuthViews.Login,
    key: 'loginView',
  },
  {
    path: '/auth/signup',
    component: AuthViews.Signup,
    key: 'signupView',
  },
  {
    redirect: true,
    path: '/auth',
    to: '/auth/login',
    key: 'loginRedirect',
  },
];

export default authRoutes;
