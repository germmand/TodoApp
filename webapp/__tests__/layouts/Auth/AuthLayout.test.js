import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { RawAuthLayout } from '../../../src/layouts/Auth/AuthLayout';

describe('Authentication Layout', () => {
  it('matches snapshot', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: { },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: () => { },
        replace: () => { },
      },
    };
    const AuthLayoutComponent = renderer.create(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    expect(AuthLayoutComponent.toJSON()).toMatchSnapshot();
  });

  it('changes route to Dashboard if user logged in on first rendering', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: true,
      location: { },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: () => { },
        replace: jest.fn(),
      },
    };
    mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    expect(props.history.replace).toHaveBeenCalledWith('/home/Dashboard');
  });

  it('changes route to Dashboard if user logs in', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: { },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: () => { },
        replace: jest.fn(),
      },
    };
    const AuthLayoutComponent = mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    const updatedProps = {
      ...props,
      isLoggedIn: true,
    };
    AuthLayoutComponent.setProps({
      children: <RawAuthLayout {...updatedProps} />,
    });
    expect(props.history.replace).toHaveBeenCalledWith('/home/Dashboard');
  });

  it('changes route to signup correctly', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: {
        pathname: '/auth/login',
      },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: jest.fn(),
        replace: () => { },
      },
    };
    const AuthLayoutComponent = mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    const SignupButton = AuthLayoutComponent.find('Button')
      .filterWhere(s => s.props().id === 'signupButton');
    SignupButton.simulate('click');
    expect(props.history.push).toHaveBeenCalledWith('/auth/signup');
  });

  it('does not change route to signup if already there', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: {
        pathname: '/auth/signup',
      },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: jest.fn(),
        replace: () => { },
      },
    };
    const AuthLayoutComponent = mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    const SignupButton = AuthLayoutComponent.find('Button')
      .filterWhere(s => s.props().id === 'signupButton');
    SignupButton.simulate('click');
    expect(props.history.push).toHaveBeenCalledTimes(0);
  });

  it('changes route to login correctly', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: {
        pathname: '/auth/signup',
      },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: jest.fn(),
        replace: () => { },
      },
    };
    const AuthLayoutComponent = mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    const LoginButton = AuthLayoutComponent.find('Button')
      .filterWhere(s => s.props().id === 'loginButton');
    LoginButton.simulate('click');
    expect(props.history.push).toHaveBeenCalledWith('/auth/login');
  });

  it('does not change route to login if already there', () => {
    const props = {
      classes: {
        grow: 'grow-auth-layout-class',
        viewContainer: 'view-container-auth-layout-class',
      },
      isLoggedIn: false,
      location: {
        pathname: '/auth/login',
      },
      history: {
        action: '',
        block: () => { },
        createHref: () => { },
        go: () => { },
        goBack: () => { },
        goForward: () => { },
        length: 1,
        listen: () => { },
        location: { },
        push: jest.fn(),
        replace: () => { },
      },
    };
    const AuthLayoutComponent = mount(
      <MemoryRouter>
        <RawAuthLayout {...props} />
      </MemoryRouter>,
    );
    const LoginButton = AuthLayoutComponent.find('Button')
      .filterWhere(s => s.props().id === 'loginButton');
    LoginButton.simulate('click');
    expect(props.history.push).toHaveBeenCalledTimes(0);
  });
});
