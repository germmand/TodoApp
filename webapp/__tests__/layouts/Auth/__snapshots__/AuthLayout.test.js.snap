// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Layout matches snapshot 1`] = `
<div>
  <header
    className="MuiPaper-root-12 MuiPaper-elevation4-18 MuiAppBar-root-3 MuiAppBar-positionStatic-7 MuiAppBar-colorPrimary-10"
  >
    <div
      className="MuiToolbar-root-39 MuiToolbar-regular-41 MuiToolbar-gutters-40"
    >
      <h6
        className="MuiTypography-root-43 MuiTypography-h6-60 MuiTypography-colorInherit-72 AuthLayout-grow-1 grow-auth-layout-class"
      >
        TodoApp - Welcome!
      </h6>
      <button
        className="MuiButtonBase-root-105 MuiButton-root-79 MuiButton-outlined-87 MuiButton-colorInherit-100"
        disabled={false}
        id="loginButton"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-80"
        >
          Login
        </span>
        <span
          className="MuiTouchRipple-root-108"
        />
      </button>
      <button
        className="MuiButtonBase-root-105 MuiButton-root-79 MuiButton-text-81 MuiButton-flat-84 MuiButton-colorInherit-100"
        disabled={false}
        id="signupButton"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-80"
        >
          Register
        </span>
        <span
          className="MuiTouchRipple-root-108"
        />
      </button>
    </div>
  </header>
  <div
    className="AuthLayout-viewContainer-2 view-container-auth-layout-class"
  />
</div>
`;
